server/src/main/java/io/druid/server/namespace/cache/OffHeapNamespaceExtractionCacheManager.java

```diff
@@ -0,0 +1,166 @@
+{
+  private static final Logger log = new Logger(OffHeapNamespaceExtractionCacheManager.class);
+  private final DB mmapDB;
+  private ConcurrentMap<String, String> currentNamespaceCache = new ConcurrentHashMap<>();
+  private Striped<Lock> nsLocks = Striped.lock(32); // Needed to make sure delete() doesn't do weird things
+
+  @JsonCreator
+  public OffHeapNamespaceExtractionCacheManager(
+      @JacksonInject Lifecycle lifecycle,
+      @JacksonInject @Named("io.druid.server.namespace.NamespacedExtractionModule")
+      ConcurrentMap<String, Function<String, String>> fnCache
+  )
+  {
+    super(lifecycle, fnCache);
+    final File tmpFile;
+    try {
+      tmpFile = File.createTempFile("druidMapDB", getClass().getCanonicalName());
+      log.info("Using file [%s] for mapDB off heap namespace cache", tmpFile.getAbsolutePath());
+    }
+    catch (IOException e) {
+      throw Throwables.propagate(e);
+    }
+    mmapDB = DBMaker
+        .newFileDB(tmpFile)
+        .closeOnJvmShutdown()
+        .transactionDisable()
+        .deleteFilesAfterClose()
+        .strictDBGet()
+        .asyncWriteEnable()
+        .mmapFileEnable()
+        .commitFileSyncDisable()
+        .cacheSize(10_000_000)
+        .make();
+    lifecycle.addHandler(
+        new Lifecycle.Handler()
+        {
+          @Override
+          public void start() throws Exception
+          {
+            // NOOP
+          }
+
+          @Override
+          public void stop()
+          {
+            mmapDB.close();
+            if(!tmpFile.delete()){
+              log.warn("Unable to delete file at [%s]", tmpFile.getAbsolutePath());
+            }
+          }
+        }
+    );
+  }
+
+  @Override
+  protected boolean swapAndClearCache(String namespaceKey, String cacheKey)
+  {
+    final Lock lock = nsLocks.get(namespaceKey);
+    lock.lock();
+    try {
+      Preconditions.checkArgument(mmapDB.exists(cacheKey), "Namespace [%s] does not exist", cacheKey);
+      final String priorCache = currentNamespaceCache.put(namespaceKey, cacheKey);
+      if (priorCache != null) {
+        mmapDB.delete(priorCache);
+        return true;
+      }
+      return false;
+    }
+    finally {
+      lock.unlock();
+    }
+  }
+
+  @Override
+  public boolean delete(final String namespaceKey)
+  {
+
+    final Lock lock = nsLocks.get(namespaceKey);
+    lock.lock();
+    try {
+      super.delete(namespaceKey);
+      final String mmapDBkey = currentNamespaceCache.get(namespaceKey);
+      if (mmapDBkey != null) {
+        mmapDB.delete(mmapDBkey);
+        return true;
+      } else {
+        return false;
+      }
+    }
+    finally {
+      lock.unlock();
+    }
+  }
+
+  @Override
+  public ConcurrentMap<String, String> getCacheMap(String cacheKey)
+  {
+    final Lock lock = nsLocks.get(cacheKey);
+    lock.lock();
+    try {
+      String realKey = currentNamespaceCache.get(cacheKey);
```

Comment:
I guess, in some cases , cacheKey could also be namespaceKey (for example in various NamespaceExtractionFnFactory.build(..) impls). it was confusing for a bit. Can we call the argument cacheOrNamespaceKey ?